ckanext/canada/plugin/security_plugin.py

```diff
@@ -72,3 +76,33 @@ def get_validators(self) -> Dict[str, Validator]:
             canada_security_upload_type=validators.canada_security_upload_type,
             canada_security_upload_presence=validators.canada_security_upload_presence,
         )
+
+    def make_middleware(self, app: CKANApp, config: 'CKANConfig') -> CKANApp:
+        return CSPNonceMiddleware(app, config)
+
+
+class CSPNonceMiddleware(object):
+    def __init__(self, app: Any, config: 'CKANConfig'):
+        self.config = config
+        self.app = app
+
+    def __call__(self, environ: Any, start_response: Any) -> Any:
+        csp_nonce = ''.join(random.choices(
+                string.ascii_letters + string.digits, k=22))
+        environ['CSP_NONCE'] = csp_nonce
+        csp_header = [
+            ('Content-Security-Policy',
+             self.config['ckanext.canada.content_security_policy'].replace(
+                 '[[NONCE]]', csp_nonce).replace('\n', ' ').replace('\r', ''))]
```

Comment:
I wonder if we want a function to clean configuration values for safe use in headers because the extra nl/cr replaces seem random

Reply:
YES. I was looking for one in Core but could not find. I will work on one here then in the canada plugin. And then can contrib upstream later?